pkg/pulumiyaml/analyser.go

```diff
@@ -1098,13 +1104,32 @@ func (tc *typeCache) typeConfig(r *Runner, node configNode) bool {
 	return true
 }
 
-// checks for config type compatibility
-func unifyConfigType(a, b schema.Type) (schema.Type, bool) {
-	a, b = codegen.UnwrapType(a), codegen.UnwrapType(b)
-	if a.String() == b.String() {
-		return a, true
-	} else if (a == schema.IntType && b == schema.NumberType) || (b == schema.IntType && a == schema.NumberType) {
+// Checks for config type compatibility between an already registered/ old config key with a new key.
+// We allow Int to fall back to Number.
+// We also allow a new non-String key to fall back to String
+// since parsing CLI-set config types is a best effort and may actually be intended as a String.
+// If the new key is a String with a defined value, and the old key is not, then we
+// parse the new key to determine if it can fall back to the old key's type
+// (i.e. "16" would be valid as an Int but "foo" would not.)
+// TODO: remove the last case once `configuration` is deprecated.
+func unifyConfigType(oldType, newType schema.Type, newVal interface{}) (schema.Type, bool) {
+	oldType, newType = codegen.UnwrapType(oldType), codegen.UnwrapType(newType)
+	if oldType.String() == newType.String() {
+		return oldType, true
+	} else if (oldType == schema.IntType && newType == schema.NumberType) || (newType == schema.IntType && oldType == schema.NumberType) {
```

Comment:
Since `unifyConfigType` is now directional, we can constrain this conversion to be one way. I'm pretty sure this will handle the use case we want: `config number` to `configuration int`.

```suggestion
	} else if newType == schema.IntType && oldType == schema.NumberType {
```


If you agree with the above change, then `unfiyConfigType(oldType, newType, newValue)` will always return `oldType, true` or `nil, false` and we can reduce its signature to `bool` (with an associated name change).

Reply:
👍🏼 Again I think it relies on the assumption that conflicts strictly arise when attempting to assign a `config` key to an already-registered `configuration` key but that should be safe